### PR TITLE
PWN-3040 - Fix additional-top-up

### PR DESCRIPTION
**Pull request by kabramovp2p**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-3040

## Description of Work

Fix additional top up

### CODE REVIEW

**Review by kuznetsov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/feerelayer/interactor/FeeRelayerInteractor.kt [hunk 1]
```diff
@@ -85,21 +85,46 @@ class FeeRelayerInteractor(
...
     suspend fun topUpAndRelayTransaction(
-        preparedTransaction: PreparedTransaction,
-        payingFeeToken: TokenInfo?
+        preparedTransactions: List<PreparedTransaction>,
+        payingFeeToken: TokenInfo,
     ): List<String> {
+        val expectedFee = FeeAmount(
+            transaction = preparedTransactions
+                .map { it.expectedFee.transaction }
+                .fold(BigInteger.ZERO, BigInteger::add),
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/feerelayer/interactor/FeeRelayerInteractor.kt:94]: maybe we can create .sum() extension for this kind of logic?

File: app/src/main/java/org/p2p/wallet/feerelayer/interactor/FeeRelayerInteractor.kt [hunk 2]
```diff
@@ -85,21 +85,46 @@ class FeeRelayerInteractor(
...
+            )
+
+            relayTransaction(
+                preparedTransaction = preparedTransactions[1]
+            )
+        } else {
+            relayTransaction(
+                preparedTransaction = preparedTransactions[0]
+            )
+        }
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/feerelayer/interactor/FeeRelayerInteractor.kt:117]:
```suggestion
       var transactionId = relayTransaction(preparedTransactions[0])
        if (preparedTransactions.size > 1) { 
            transactionId = relayTransaction(preparedTransactions[1])
        }
```

File: app/src/main/java/org/p2p/wallet/feerelayer/interactor/FeeRelayerSwapInteractor.kt [hunk 3]
```diff
@@ -129,6 +130,11 @@ class FeeRelayerSwapInteractor(
             expectedFee.transaction += lamportsPerSignature
         }
 
+        // in transitive swap, there will be situation when swapping from SOL -> SPL that needs spliting transaction to 2 transactions
+        if (swapPools?.size == 2 && sourceTokenMint == WRAPPED_SOL_MINT && destinationAddress == null) {
+            expectedFee.transaction += lamportsPerSignature * BigInteger.valueOf(2L)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/feerelayer/interactor/FeeRelayerSwapInteractor.kt:135]: 2L should be extracted as const

File: app/src/main/java/org/p2p/wallet/feerelayer/interactor/FeeRelayerSwapInteractor.kt [hunk 4]
```diff
@@ -462,4 +489,28 @@ class FeeRelayerSwapInteractor(
...
+
+    private fun prepareTransaction(
+        instructions: List<TransactionInstruction>,
+        signers: List<Account>,
+        blockhash: String,
+        feePayerAddress: PublicKey,
+        accountCreationFee: BigInteger,
+        lamportsPerSignature: BigInteger
+    ): PreparedTransaction {
+        val transaction = Transaction()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/feerelayer/interactor/FeeRelayerSwapInteractor.kt:501]: apply?

File: app/src/main/java/org/p2p/wallet/send/interactor/SendInteractor.kt [hunk 5]
```diff
@@ -156,18 +157,18 @@ class SendInteractor(
...
             transactionInteractor.serializeAndSend(
                 preparedTransaction = preparedTransaction,
                 isSimulation = false
             )
+        } else {
+            // use fee relayer
+            feeRelayerInteractor.topUpAndRelayTransaction(
+                preparedTransactions = listOf(preparedTransaction),
+                payingFeeToken = TokenInfo(feePayerToken.publicKey, feePayerToken.mintAddress)
+            ).firstOrNull().orEmpty()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/send/interactor/SendInteractor.kt:171]:
```suggestion
            )
            .firstOrNull()
            .orEmpty()
```

**Review by gslevinkov — APPROVED** [review]

**Pull request closed without merging**